tools/batch-script/address_provider.go

```diff
@@ -0,0 +1,224 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2022 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package batch_script
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go-sdk/client"
+	"github.com/rs/zerolog"
+)
+
+// AddressProvider Is used to get all the addresses that exists at a certain referenceBlockId
+type AddressProvider struct {
+	log              zerolog.Logger
+	lastAddress      flow.Address
+	generator        *flow.AddressGenerator
+	lastAddressIndex uint
+	referenceBlockID flow.Identifier
+	currentIndex     uint
+}
+
+const endOfAccountsError = "storage used is not initialized or not initialized correctly"
+
+const accountStorageUsageScript = `
+pub fun main(address: Address): UInt64 {
+  return getAccount(address).storageUsed
+}
+`
+
+// InitAddressProvider uses bisection to get the last existing address.
+func InitAddressProvider(
+	ctx context.Context,
+	log zerolog.Logger,
+	chain flow.ChainID,
+	referenceBlockID flow.Identifier,
+	client *client.Client,
+	pause time.Duration,
+) (*AddressProvider, error) {
+	ap := &AddressProvider{
+		log:              log,
+		generator:        flow.NewAddressGenerator(chain),
+		referenceBlockID: referenceBlockID,
+		currentIndex:     1,
+	}
+
+	last := time.Now()
+
+	searchStep := 0
+	addressExistsAtIndex := func(index uint) (bool, error) {
+		if time.Since(last) < pause {
+			time.Sleep(pause)
+		}
+		last = time.Now()
+
+		searchStep += 1
+		address := ap.indexToAddress(index)
+
+		// This script will fail with endOfAccountsError
+		// if the account (address at given index) doesn't exist yet
+		_, err := client.ExecuteScriptAtBlockID(
+			ctx,
+			referenceBlockID,
+			[]byte(accountStorageUsageScript),
+			[]cadence.Value{cadence.NewAddress(address)},
+		)
+		if err == nil {
+			return true, nil
+		}
+		if strings.Contains(err.Error(), endOfAccountsError) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	// We assume address #2 exists
+	lastAddressIndex, err := ap.getLastAddress(1, 2, true, addressExistsAtIndex)
```

Comment:
Maybe short-circuit the check if address `#2` doesn't exist?

Reply:
I just copied the internal code as is and didn't notice this. @janezpodhostnik Do you know why address 2 is assumed to exist? (maybe you @tehranifar?)

Reply:
Seems like the code was refactored by @janezpodhostnik and @farbodg
The original version was meant to drive a script that operates on a range of accounts, e.g. `30, 31, 32`. If one of those accounts doesn't exist the script panics so the code had to try with `30, 31` and then `30` to make sure it's capturing the end range correctly. The new version is probably achieving the same.

Reply:
I added that assumption (that account 2 exists) because all the networks the program would be run on had account 2.

During bootstrapping we create at least 4 accounts.

If this is a problematic assumption, the code just needs to test that address no. 2 exists, and if it doesn't address no. 1 is the last one.